fedimint-core/src/encoding/mod.rs

```diff
@@ -379,6 +379,29 @@ impl Encodable for lightning_invoice::Invoice {
     }
 }
 
+impl Encodable for lightning::routing::gossip::RoutingFees {
```

Comment:
Not sure if it's a good idea to depend for our own encodings on data-structures from external crates (that possibly could change). 

Also how we are going to implement gateway fees and LN gateway fees, and probably two different things. 

We might have fees paid in different currencies supported by federation, on maybe "no gateway fee tuesdays". That's why I think we should go with an enum. But the main variant of that enum should probably be this thing. Not sure if we should copy&paste or use directly.

Reply:
> RE: Also how we are going to implement gateway fees and LN gateway fees, are probably two different things.
>
> Could have fees paid in different currencies supported by federation, or maybe "no gateway fee tuesdays", or some other things? If not, then whatever, let's just go with this, seems reasonable. If possibly yes, maybe we should go with an enum. But the main variant of that enum should probably be this thing. Not sure if we should copy&paste or use directly.

Fantastic idea IMO. If I adopt it as a proposal, we'd have
```rs
#[non_exhaustive]
pub enum FeeStructure {
  LnRouting { fees: RoutingFees }
}
```
To open up future possibility of:
```rs
#[non_exhaustive]
pub enum FeeStructure {
  LnRouting { fees: RoutingFees },
  Discount { ... }
}
```
Because "Free Gateway Tuesdays" is a fantastic thing :)

Reply:
> RE: Not sure if it's a good idea to depend for our own encodings on data-structures from external crates (that possibly could change).

We are also [guilty in these other accounts,](https://github.com/fedimint/fedimint/blob/master/fedimint-core/src/encoding/mod.rs#L376-L391) and I'm not keen on maintaining a variant of [RoutingFees](https://github.com/lightningdevkit/rust-lightning/blob/main/lightning/src/routing/gossip.rs#L1057-L1065) because we already have a dependency on the crate and type in other scenarios.

Can we fully rely on version control to alert breaks in encodings?  Is maintaining our own types the only way to avoid encoding extern types?

Reply:
Proposing flexible fee structure as a quick follow up to the current PR: #2287

Reply:
One way I could also see this working in the future is that clients present the gateway(s) with an invoice _before_ they create the outgoing contract essentially asking them what the fees would be to pay this invoice. Default behavior could be just computing a fee based fixed `base_msat` / `proportional_millionths`. But more sophisticated gateways could probe and guess which channel they would pay across in case they had different channels with different routing fees. So fees would be a function of the invoice being paid.